### DISCUSSION TITLE
Implement synchronous toggle function in details page

### DISCUSSION THREAD

**Discussion by MeenaFoerderer** [neuefische/questions#364] [Web] [answered]
Hello,
I have an array of objects that is rendered dynamically on the [date]/[room] index.js page. I was able to implement toggle functionality via a bookmark icon, also so that bookmarked objects are saved to local storage. I also have a details [id] page. I would like the toggle functionality to work synchronously on the detail page as well, but I am having trouble getting that to work. Help is very much appreciated.

my PR: https://github.com/MeenaFoerderer/Capstone-project/pull/18
error message: 
![Bildschirmfoto 2022-12-03 um 14 40 38](https://user-images.githubusercontent.com/105375266/205444083-ea65b2ad-7209-4e3b-9610-8f0e2c908ade.png)

**Comment by JessicaLoers** [comment, marked as answer]
**Hi @MeenaFoerderer** 

The error message from the screenshot is caused because `talk` is not defined -  it should be `talks` (see line 13)
**BUT**
You already "unpacked" the talks object by destructuring assignment (line 17 - 25). Simply add the `isBookmarked` key.
```js
 const {
    // ...
    isBookmarked
  } = talks.find((talk) => talk.id === id);
```

**I would also like to recommend the following**

1.  place the toggle function (which handles the talks state change) inside the `_app.js`. Keep in mind:  never pass the state-setter (callback function  e.g. `setTalks`) to your components 🚫
2. rename the `toggleFavourite` function to `handleBookmarkToggle` and pass this function as a prop `onBookmarkToggle={handleBookmarkToggle}`

**_app.js**
```js
// ...

 function handleBookmarkToggle(id) {
    const updatedTalks = talks.map((talk) =>
      talk.id === id ? { ...talk, isBookmarked: !talk.isBookmarked } : talk
    );
    setTalks(updatedTalks);
  }
return (
    <>
      <GlobalStyles />
      <Component
        {...pageProps}
        conferenceDays={conferenceDays}
        conferenceRooms={conferenceRooms}
        talks={talks}
        onBookmarkToggle={handleBookmarkToggle}
      />
    </>
  );
```
**talks/[id].js**
```js
function TalkDetails({ talks, onBookmarkToggle }) {
//...

return (
//...
          <Button
            onClick={() => {
              onBookmarkToggle(id);
            }}>
            {isBookmarked ? <BookmarkActive /> : <BookmarkInactive />}
          </Button>
//...
 )
}
```
‼️ **Important note: The code above is written "dry", so please don't copy and paste.**

**Reply by MeenaFoerderer** [reply]
Hi @JessicaLoers 
thanks very much for your reply. I implemented everything as suggested, however, the toggle functionality in the `talks/[id].js` is still not working. 
I now added the isBookmarked key but it is apparently not used in the code. I also changed `talk` in lines 46 and 49 to `talks` but that doesn't do anything either. It just removes the error message. However, the `talks` also should not have an id as it is the `data`, i.e. the entire array. I need to refer to the specific `talk.id` but somehow `talk` is not recognized in the `onClick` codeblock (from line 45 on) and I don't know how to change that.

My updated PR: https://github.com/MeenaFoerderer/Capstone-project/pull/18

**Comment by JessicaLoers** [comment]
Hi @MeenaFoerderer 

**I left you some suggestions in the PR - that's certainly a bit easier for you than writing it down here.** 😬
There are a few lines that you should adjust.  [PR LINK](https://github.com/MeenaFoerderer/Capstone-project/pull/18#pullrequestreview-1203660321)

I tested the code on my local machine and the bookmark functionality works well.

**Reply by MeenaFoerderer** [reply]
Thanks a lot, Jessica! It was very helpful.
